### DISCUSSION TITLE
How do I figure out why binary caching doesn't restore packages?

### DISCUSSION THREAD

**Discussion by niyue** [microsoft/vcpkg#20569] [Q&A] [answered]
I start to use vcpkg binary caching feature to avoid unnecessary re-compilation. I setup env var `VCPKG_BINARY_SOURCES=clear;files,/vcpkg/archives,readwrite` in Dockerfile for configuration. 

For some reason, this setup works when building the Docker image, and I can see outputs like `Restored 168 packages from /vcpkg/archives in 8.181 s. Use --debug to see more details.`, however, when I uses the built Docker image in CI, it reports `Restored 0 packages from /vcpkg/archives` and triggered a re-build. 

I tried using the `--debug` switch to get more information, but currently, it only reports several hash about compiler/cmake/port files. For example,
```
 [DEBUG] <abientries for libuuid:x64-linux>
 [DEBUG]   CMakeLists.txt|eed0900ba63f326fc9be4d88cc9b7ab9d53e70005303dfd052e0fbf3e772cc09
 [DEBUG]   cmake|3.21.1
 [DEBUG]   config.linux.h|95b82bd74377635a7e67a95b1b96143b370443abc054fc42c728e5d5a4153297
 [DEBUG]   features|core
 [DEBUG]   portfile.cmake|cbbc2da7432eb7dc887edf46fa03d4700ecd2934cef712494b3d6418723b2bf7
 [DEBUG]   ports.cmake|4622bd68add4d50fc08fde4a5d1faf937493bd54f1e5352ca2242a92dc83fa04
 [DEBUG]   post_build_checks|2
 [DEBUG]   triplet|x64-linux
 [DEBUG]   triplet_abi|561f6d71d6bcccbee8feaecf7eedab6f2f77370399ba1fcad0507b6dc62def58-dbf0dd3cf61bd48347b4c28cd439488aeb7a6f756ba96cca5889a45c1972b236-8a43818be07c62b3462dfba4c50034ab968903d7
 [DEBUG]   unofficial-libuuid-config.cmake.in|d440b00400bed20ebc6c0e4ef9f646c558b8e752fa11f0a78ff0392a0653f584
 [DEBUG]   vcpkg.json|2fee1ea5c459460c259fdf2e6cc46d46c76184a3f41609456ef42a4e6963e479
 [DEBUG]   vcpkg_configure_cmake|431fc1a744f00f7a8f36de8d4758d283210d7fc36b8f62009a0f1a160b2ee7f6
 [DEBUG]   vcpkg_copy_pdbs|59e4c0d2321697848a899ba9537394cdee81725b11965c6d1341df53a6850380
 [DEBUG]   vcpkg_fail_port_install|01b43acb1a7a8903f82a0a6fe9c005bef12e7f2189f78db317c60628fa457384
 [DEBUG]   vcpkg_fixup_cmake_targets|b85183aa9f7bb5f84a9f57310eb81cb8c1744008f786e7ecb27becfa0bf9b05c
 [DEBUG]   vcpkg_fixup_pkgconfig|ff54f8b06c83b54a1af5e35286a6c36ea9e69376bed1d11b5521c254987c123f
 [DEBUG]   vcpkg_from_sourceforge|1d9abe7894920ee65041c6bf92cfd158f905bad7e483b6c4f1185e866fcefb2e
 [DEBUG]   vcpkg_install_cmake|a04ca69ec49d413d16091484556ec2c9429a4e1e002cbfd2fc20bb24644aa88c
 [DEBUG] </abientries>
```

I assume they are for the latest port definition obtained, but how do I know **why** my existing vcpkg binary cache (in my case, the files in `/vcpkg/archives` folder) doesn't match the above information and why the re-build is needed?

**Comment by niyue** [comment, marked as answer]
I figured out the problem for my case. 
1) For each installed port, you will be able to see all its hash under the cache folder. 
  1.1) unzip a cached port, /vcpkg/archives/01/018efeb9fd88aad2c9fcd34d64724cbb17fc24b587ec491fedad8aedb0ea3e59.zip
   `unzip /vcpkg/archives/01/018efeb9fd88aad2c9fcd34d64724cbb17fc24b587ec491fedad8aedb0ea3e59.zip -d myport`
  1.2) there will be a `vcpkg_abi_info.txt` in the archive. it contains all the hash info. 
  1.3) You should check out the existing cached version of `vcpkg_abi_info.txt` and the new port's hash values, and compare them to see which hash differs.

2) In my case, the `triplet_abi` hash differs. It has several parts, and the last part is compiler hash. Compiler hash will be shown when `--debug` flag is turned on when installing. 

3) So I need to figure out why compiler hash differs in my env. The actual compiler used for generating the compiler hash is shown when `--debug` flag is turned on. In my case, one env uses `/usr/lib/ccache/g++` and the other env uses `/usr/bin/g++`, so they generate different compiler hash. The ccache is installed in the Docker image initially, so both of them have ccache. I find cmake to detect different compiler because one Docker image uses `/bin/sh` and the other use `/bin/bash` as the shell, and the latter happens to use a `.bashrc` adding ccache into the `PATH` env var, so cmake detects the ccache as the compiler in this env.

**Reply by Osyotr** [reply]
Note that you can also use `VCPKG_DISABLE_COMPILER_TRACKING` to disable compiler tracking.
